File: Battleships/Player.cpp
#include "Player.hpp"


int Player::playerCount = 0;

Player::Player(const std::string& name) : name(name) {
    playerCount++;
    statistics["hits"] = 0;
    statistics["misses"] = 0;
    statistics["ships_destroyed"] = 0;
}

bool Player::allShipsSunk() const {
    return std::all_of(ships.begin(), ships.end(),
        [](const Ship& ship) { return ship.isSunk(); });
}

ShotResult Player::getShotResult(const Coordinate& coord) {
    ShotResult result = myBoard.receiveShot(coord);

    if (result == ShotResult::Hit || result == ShotResult::Sunk) {
        auto it = std::find_if(ships.begin(), ships.end(),
            [&coord](const Ship& ship) {
                const auto& coords = ship.getCoordinates();
                return std::any_of(coords.begin(), coords.end(),
                    [&coord](const Coordinate& c) { return c == coord; });
            });

        if (it != ships.end()) {
            bool wasSunk = it->takeHit(coord);
            if (wasSunk) {
                result = ShotResult::Sunk;
                myBoard.markAreaAroundSunkShip(*it);
                statistics["ships_destroyed"]++;
            }
        }
    }

    return result;
}

void Player::addShip(const Ship& ship) {
    ships.push_back(ship);
}

void Player::updateStatistics(const std::string& key, int value) {
    statistics[key] += value;
}

File: SeaBattleCSharp/Player.cs
using System.Collections.Generic;
using System.Linq;

namespace SeaBattleCSharp
{
    public abstract class Player
    {
        protected string name;
        protected GameBoard myBoard;
        protected GameBoard enemyBoard;
        protected List<Ship> ships;
        protected Dictionary<string, int> statistics;

        public static int PlayerCount { get; private set; }

        public Player(string name)
        {
            this.name = name;
            myBoard = new GameBoard();
            enemyBoard = new GameBoard();
            ships = new List<Ship>();
            statistics = new Dictionary<string, int>
            {
                ["hits"] = 0,
                ["misses"] = 0,
                ["ships_destroyed"] = 0
            };
            PlayerCount++;
        }

        public abstract bool IsHuman();
        public virtual bool IsAdvancedAI() => false;
        public abstract bool MakeMoveWithResult(Player enemy);
        public abstract void PlaceShips();
        public abstract void MakeMove(Player enemy);
        public abstract void MarkAreaAroundDestroyedShip(Player enemy, Coordinate hitCoord);
        public virtual string GetPlayerType() => "Base Player";

        public bool AllShipsSunk()
        {
            return ships.All(ship => ship.IsSunk());
        }

        public ShotResult GetShotResult(Coordinate coord)
        {
            ShotResult result = myBoard.ReceiveShot(coord);

            if (result == ShotResult.Hit || result == ShotResult.Sunk)
            {
                var ship = ships.FirstOrDefault(s =>
                    s.Coordinates.Any(c => c.Equals(coord)));

                if (ship != null)
                {
                    bool wasSunk = ship.TakeHit(coord);
                    if (wasSunk)
                    {
                        result = ShotResult.Sunk;
                        myBoard.MarkAreaAroundSunkShip(ship);
                        statistics["ships_destroyed"]++;
                    }
                }
            }
            return result;
        }

        public string GetName() => name;
        public GameBoard GetMyBoard() => myBoard;
        public GameBoard GetEnemyBoard() => enemyBoard;
        public List<Ship> GetShips() => ships;

        public void AddShip(Ship ship)
        {
            ships.Add(ship);
        }

        public void UpdateStatistics(string key, int value)
        {
            if (statistics.ContainsKey(key))
                statistics[key] += value;
        }
    }
}